torch/csrc/dynamo/guards.cpp

```diff
@@ -1676,14 +1699,274 @@ class RootGuardManager : public GuardManager {
+  // the key and value managers.
+  std::vector<GuardAccessor*> get_accessors() const override {
+    throw std::runtime_error(
+        "KeyValueDictGuardManager does not have accessors");
+  }
+
+  // Debug helper - Returning raw pointers because we can't return unique_ptr
+  // and pybind does not accept a unique_ptr reference return type.
+  virtual std::vector<GuardManager*> get_child_managers() override {
+    std::vector<GuardManager*> ret;
+    ret.push_back(_key_manager.get());
+    ret.push_back(_value_manager.get());
+    return ret;
+  }
+
+ private:
+  bool _is_dict_guard_manager() override {
+    return true;
+  }
+
+ private:
+  std::unique_ptr<GuardManager> _key_manager;
+  std::unique_ptr<GuardManager> _value_manager;
+};
+
+/*
+ * Dicts are common in python code. Therefore, we handle guards for dicts
+ * differently and use PyDict_* APIs which are faster than PyObject_* APIs
+ * because of no ref count increments/decrements.
+ *
+ * DictGuardManager is composed of a vector of KeyValueDictGuardManager - an
+ * encapsulation of a manager for (key, value) pair. This ensures that key guard
+ * managers are always run before value guard managers.
+ *
+ * DictGuardManager relies on the order of dict.keys(). It keeps track of the
+ * indices of dict.keys() to access the key, value pair.
+ */
+class DictGuardManager : public GuardManager {
+ public:
+  DictGuardManager(RootGuardManager* root, py::handle example_value)
+      : GuardManager(root), _size(PyDict_Size(example_value.ptr())) {}
+
+  /**
+   * Adds a new KeyDictGuardAccessor. If the accessor is already present, we
+   * just return the guard manager.
+   */
+  virtual GuardManager* get_key_value_manager(
+      const py::object& accessor_key) override {
+    // Check if the accessor is already present.
+    Py_ssize_t index = py::cast<Py_ssize_t>(accessor_key);
+    auto it = _key_value_managers.find(index);
+    if (it != _key_value_managers.end()) {
+      return it->second.get();
+    }
+    _indices.push_back(index);
+    // Always keep the _indices array sorted
+    std::sort(_indices.begin(), _indices.end());
+    _key_value_managers[index] =
+        std::make_unique<KeyValueDictGuardManager>(this->get_root());
+    return _key_value_managers[index].get();
+  }
+
+  virtual bool check_nopybind(PyObject* obj) override { // borrowed ref
+    // TODO(janimesh) - Implement a fast-path using dict versions.
+
+    if (!PyDict_Check(obj)) {
+      _fail_count += 1;
+      return false;
+    }
+
+    if (PyDict_Size(obj) != _size) {
+      _fail_count += 1;
+      return false;
+    }
+
+    PyObject *key, *value;
+    Py_ssize_t pos = 0;
+
+    // Points to an element in the _indices vector.
+    size_t index_pointer = 0;
+
+    while (index_pointer < _indices.size() &&
+           PyDict_Next(obj, &pos, &key, &value)) {
+      // Pos moves with PyDict_Next
+      Py_ssize_t pos_minus_1 = pos - 1;
```

Comment:
I don't think `pos` is guaranteed to be the same across two dicts.  It is some internal state thing, and has no meaningful use here.

You should just use your own counter and increment it each iteration.

Reply:
Oh thanks for catching this. Debugging this would have been nightmare.

Updated.